File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.py
def solution(places):
    def dfs(depth, r, c):
        if depth == 2:  # 최대 거리 2까지만 탐색
            return True

        visited[r][c] = 1

        for i in range(4):
            nr = r + dr[i]
            nc = c + dc[i]

            if 0 <= nr < 5 and 0 <= nc < 5 and not visited[nr][nc]:
                if place[nr][nc] == 'P':  # 거리 2 이내에 P가 있으면 거리두기 실패
                    return False
                if place[nr][nc] == 'O':  # 빈 공간일 경우만 계속 탐색
                    if not dfs(depth + 1, nr, nc):
                        return False

        return True

    answer = []
    dr = [-1, 1, 0, 0]
    dc = [0, 0, -1, 1]

    for place in places:
        flag = True  
        for i in range(5):
            for j in range(5):
                if place[i][j] == 'P':  
                    visited = [[0] * 5 for _ in range(5)]
                    if not dfs(0, i, j):  
                        flag = False
                        break
            if not flag:
                break
        answer.append(1 if flag else 0)

    return answer

File: 프로그래머스/2/81302. 거리두기 확인하기/거리두기 확인하기.java
class Solution {
    static String[] map = new String[5];
    // 좌,하,우
    static int[] dr = {0,1,0};
    static int[] dc = {-1,0,1};
    static int[] answer = new int [5];
    static int [][] visited=new int[5][5];
    
    public int[] solution(String[][] places) {
        for (int i=0;i<5;i++){
            map = places[i];
            visited=new int[5][5];
            if(isRightDistance()){
                answer[i]=1;
            }
        }
        return answer;
    }
    
    
    static boolean isRightDistance(){
        for (int i=0;i<5;i++){
            for(int j=0;j<5;j++){
                if (map[i].charAt(j)=='P'){
                    visited[i][j]=1;
                    if(check(i,j)){
                        for (int k = 0; k<3;k++){
                            int nr = i+dr[k];
                            int nc = j+dc[k];
                            
                            if( nr>=0 && nr<5 && nc>=0 && nc<5 && visited[nr][nc]==0) {
                                if(!check(nr,nc))return false;
                            }
                            
                    }    
                    } else {
                        return false;
                    }
                }
                    
            }
        }
        return true;
    }
    static boolean check(int i, int j){
        if (map[i].charAt(j)=='X') return true;
        for (int k = 0; k<3;k++){
            int nr = i+dr[k];
            int nc = j+dc[k];
            
            if( nr>=0 && nr<5 && nc>=0 && nc<5){
                if (visited[nr][nc]==0 && map[nr].charAt(nc)=='P') return false;
            }
        }
        return true;
        
    }
}